src/unit_tests/os_auth/test_auth_validate.c

```diff
@@ -0,0 +1,344 @@
+    snprintf(merror_message, 2048, "Invalid agent name %s (same as manager)", host_name);  
+    expect_string(__wrap__merror, formatted_msg, merror_message);        
+    err = w_auth_validate_data(response,NEW_IP1, host_name, NULL);
+    assert_int_equal(err, OS_INVALID);
+    assert_string_equal(response, err_response);  
+
+    /* Check no agent was deleted*/
+    assert_true(keys.keysize == 3);
+    int index = 0;
+    index = OS_IsAllowedName(&keys, EXISTENT_AGENT1);
+    assert_true(index >= 0);
+    index = OS_IsAllowedName(&keys, EXISTENT_AGENT2);
+    assert_true(index >= 0);
+}
+
+static void test_w_auth_validate_data_force_insert(void **state) {    
+
+    char response[2048] = {0};   
+    w_err_t err;
+
+    /* Duplicated IP*/
+    response[0] = '\0';
+    expect_string(__wrap__minfo, formatted_msg, "Duplicated IP '"EXISTENT_IP1"' (001). Saving backup.");            
+    err = w_auth_validate_data(response, EXISTENT_IP1, NEW_AGENT1, NULL);  
+    assert_int_equal(err, OS_SUCCESS);
+    assert_string_equal(response, "");  
+
+     /* Duplicated Name*/
+    response[0] = '\0';
+    expect_string(__wrap__minfo, formatted_msg, "Duplicated name '"EXISTENT_AGENT2"' (002). Saving backup.");
+    err = w_auth_validate_data(response, NEW_IP2, EXISTENT_AGENT2, NULL);  
+    assert_int_equal(err, OS_SUCCESS);
+    assert_string_equal(response, "");    
+
+    /* Check agents were deleted*/
+    int index = 0;
+    index = OS_IsAllowedIP(&keys, EXISTENT_IP1);
+    assert_true(index < 0);
+    index = OS_IsAllowedName(&keys, EXISTENT_AGENT2);
+    assert_true(index < 0);  
+}
+
+static void test_w_auth_validate_data_register_limit(void **state) {
+    char response[2048] = {0};   
+    char agent_name[2048];
+    char error_message[2048];
+    w_err_t err;
+
+    
+    //Filling most of keys element with a fixed key to reduce computing time
+    char fixed_key[KEYSIZE] = "1234";    
+    for(unsigned i=0; i<MAX_AGENTS-10; i++) {
+        OS_AddNewAgent(&keys, NULL, agent_name, ANY_IP, fixed_key);
+    }
+    
+    //Adding last keys as usual
+    for(unsigned i=0; i<10; i++) {
+        snprintf(agent_name, 2048, "__agent_%d", i);
+        response[0] = '\0';  
+        if(keys.keysize >= (MAX_AGENTS - 2)) {
+            snprintf(error_message, 2048, AG_MAX_ERROR, MAX_AGENTS - 2);
+            expect_string(__wrap__merror, formatted_msg, error_message);
+            err = w_auth_validate_data(response,ANY_IP, agent_name, NULL);            
+            assert_int_equal(err, OS_INVALID);
+            assert_string_equal(response, "ERROR: The maximum number of agents has been reached\n\n");  
+        }
+        else {
+            err = w_auth_validate_data(response,ANY_IP, agent_name, NULL);            
+            assert_int_equal(err, OS_SUCCESS);
+            assert_string_equal(response, "");  
+        }
+        OS_AddNewAgent(&keys, NULL, agent_name, ANY_IP, NULL);
+    }
+}
+
+static void test_w_auth_validate_groups(void **state) {
```

Comment:
This method takes a long time to run on Valgrind. Probably because it is adding the agent keys for real. If we are only unit testing all external funcionality should be mocked.

Reply:
Valgrind is taking lot of time in test_w_auth_validate_data_register_limit.
keys store should be mocked but is not possible with actual architecture of keys module, because every search on the store implies to receive an index instead of an element. This makes that authd handles the keys directly and no mocking is possible.
On a future design of keystores, we can separate each functionality and mock them.

Reply:
Noted, I haven´t taken into account the keystore problems. We'll leave this as it is then! Thanks for the explanation